packages/js/components/src/image-gallery/image-gallery.tsx

```diff
@@ -1,33 +1,164 @@
 /**
  * External dependencies
  */
-import { Children, createElement } from '@wordpress/element';
+import {
+	createElement,
+	cloneElement,
+	useState,
+	useEffect,
+	useCallback,
+} from '@wordpress/element';
+import classnames from 'classnames';
+import { MediaUpload } from '@wordpress/media-utils';
 
 /**
  * Internal dependencies
  */
-import { Sortable, SortableHandle } from '../sortable';
+import { Sortable, moveIndex } from '../sortable';
+import { ImageGalleryToolbar } from './index';
+import { ImageGalleryChild, MediaUploadComponentType } from './types';
 
 export type ImageGalleryProps = {
-	children: JSX.Element | JSX.Element[];
+	children: ImageGalleryChild | ImageGalleryChild[];
 	columns?: number;
+	onRemove?: ( props: {
+		removeIndex: number;
+		removedItem: ImageGalleryChild;
+	} ) => void;
+	onReplace?: ( props: {
+		replaceIndex: number;
+		previousItem: ImageGalleryChild;
+		newItem: ImageGalleryChild;
+	} ) => void;
+	onOrderChange?: ( items: ImageGalleryChild[] ) => void;
+	MediaUploadComponent?: MediaUploadComponentType;
 } & React.HTMLAttributes< HTMLDivElement >;
 
 export const ImageGallery: React.FC< ImageGalleryProps > = ( {
 	children,
 	columns = 4,
+	onOrderChange = () => null,
+	onRemove = () => null,
+	onReplace = () => null,
+	MediaUploadComponent = MediaUpload,
 }: ImageGalleryProps ) => {
+	const [ toolBarItem, setToolBarItem ] = useState< string | null >( null );
+	const [ orderedChildren, setOrderedChildren ] = useState<
+		ImageGalleryChild[]
+	>( [] );
+
+	useEffect( () => {
+		if ( ! children ) {
+			return;
+		}
+		setOrderedChildren(
+			( Array.isArray( children ) ? children : [ children ] ).map(
+				( child, index ) =>
+					cloneElement( child, { key: child.key || String( index ) } )
+			)
+		);
+	}, [ children ] );
+
+	const updateOrderedChildren = ( items: ImageGalleryChild[] ) => {
+		setOrderedChildren( items );
+		onOrderChange( orderedChildren );
+	};
+
+	const moveItem = useCallback(
```

Comment:
It would be really great to have `moveItem`, `removeItem`, and `replaceItem` extracted to utils with some unit tests.

The state methods would need to be replaced with callbacks in order for this to work.

Reply:
`moveItem` doesn't add any logic to `moveIndex`, so I've left that one out. I've moved the others into the `utils.ts` file and I'm adding tests now.